optuna/storages/_heartbeat.py

```diff
@@ -25,13 +101,22 @@ def fail_stale_trials(study: "optuna.Study") -> None:
     if not storage.is_heartbeat_enabled():
         return
 
+    heartbeat = get_heartbeat(storage)
```

Comment:
Not sure but how about using TypeGuard here and removing `NullHeartbeat` class? The code is like below:

```python
try:
    from typing import TypeGuard
except ImportError:
    from typing_extensions import TypeGuard


def is_heartbeat_enabled(storage: BaseStorage)->TypeGuard[HeartbeatStorage]:
    if isinstance(storage, BaseHeartbeat) and self.get_heartbeat_interval() is not None:
        return True
    return False
```

```python
def fail_stale_trials(study: Study) -> None:
    storage = study._storage
    if not is_heart_beat_enabled(storage):
        return

    failed_trial_ids = []
    for trial_id in storage._get_stale_trial_ids(study._study_id):
        ...
```

Links:

* https://docs.python.org/ja/3/library/typing.html#typing.TypeGuard
* https://github.com/python/typing/tree/master/typing_extensions